pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,355 @@
+	pool *connpool
+	host types.CreateConnectionData
+
+	keepAlive *keepAliveListener
+
+	reconnectState uint64 // for reconnect, connecting or notConnecting
+}
+
+func (p *connpool) newActiveClient(ctx context.Context) (*activeClient, types.PoolFailureReason) {
+	// if pool is already destroyed, return
+	if atomic.LoadUint64(&p.destroyed) == 1 {
+		return nil, types.ConnectionFailure
+	}
+
+	ac := &activeClient{
+		pool: p,
+		host: p.Host().CreateConnection(ctx),
+	}
+
+	// this should be equal to
+	// codecClient.AddConnectionEventListener(ac) // ac.OnEvent
+	ac.host.Connection.AddConnectionEventListener(ac)
+
+	// first connect to dest addr, then create stream client
+	if err := ac.host.Connection.Connect(); err != nil {
+		if p.autoReconnectWhenClose && atomic.LoadUint64(&p.destroyed) == 0 {
+			// auto reconnect when the first connect failed
+			ac.Reconnect()
+		}
+
+		return nil, types.ConnectionFailure
+	} else {
+		if atomic.LoadUint64(&p.destroyed) == 1 {
+			// if destroyed, close the conn
+			ac.host.Connection.Close(api.NoFlush, api.LocalClose)
+			return nil, types.ConnectionFailure
+		}
+
+		for _, rf := range ac.pool.readFilters {
+			ac.host.Connection.FilterManager().AddReadFilter(rf)
+		}
+	}
+
+	// if user use connection pool without codec
+	// they can use the connection returned from activeClient.Conn()
+	if p.heartBeatCreator != nil {
+		ac.SetHeartBeater(p.heartBeatCreator())
+	}
+
+	return ac, ""
+}
+
+// Reconnect triggers connection to reconnect
+func (ac *activeClient) Reconnect() {
+	if !atomic.CompareAndSwapUint64(&ac.reconnectState, notConnecting, connecting) {
+		return
+	}
+
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		return
+	}
+
+	// TODO, remove this
+	println(string(debug.Stack()))
+
+	utils.GoWithRecover(func() {
+		defer atomic.CompareAndSwapUint64(&ac.reconnectState, connecting, notConnecting)
+		var (
+			err error
+			i   int
+		)
+
+		// close previous conn
+		ac.host.Connection.Close(api.NoFlush, api.RemoteClose)
+
+		var (
+			backoffArr = []time.Duration{
+				time.Second,
+				time.Second * 2,
+				time.Second * 5,
+				time.Second * 10,
+			}
+			backoffIdx = 0
+		)
+
+		for ; i < ac.pool.reconnTryTimes; i++ {
+			if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+				// if pool was exited, then stop
+				return
+			}
+			// build new conn
+			// must create this new conn, the same conn can only be connected once
+			ac.host = ac.pool.Host().CreateConnection(context.Background())
+
+			// connect the new connection
+			err = ac.host.Connection.Connect()
+			if err != nil {
+				log.DefaultLogger.Warnf("[connpool] reconnect failed %v times, host %v", i+1, ac.host.Host.AddressString())
+
+				// backoff logic
+				if backoffIdx >= len(backoffArr) {
+					backoffIdx = len(backoffArr) - 1
+				}
+
+				time.Sleep(backoffArr[backoffIdx])
+				backoffIdx++
+				continue
+			}
+
+			log.DefaultLogger.Infof("[connpool] reconnect succeed after %v tries, host %v", i+1, ac.host.Host.AddressString())
```

Comment:
it looks good,but logging on connected/disconnected event in OnEvent function of activeClient might be better?

Reply:
OnEvent also have logs to record the cause for reconnecting

```go
	switch {
	case event.IsClose():
		// RemoteClose when read/write error
		// LocalClose when there is a panic
		// OnReadErrClose when read failed
		log.DefaultLogger.Warnf("[connpool] reconnect after conn close, host : %v", ac.host.Host.AddressString())
		goto RECONN

	case event == api.ConnectTimeout:
		log.DefaultLogger.Warnf("[connpool] connect timeout, host : %v", ac.host.Host.AddressString())
		goto RECONN
	case event == api.ConnectFailed:
		log.DefaultLogger.Warnf("[connpool] connect failed, host : %v", ac.host.Host.AddressString())
		goto RECONN
	}
```